pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/SchemaChangeHandler.java

```diff
@@ -0,0 +1,341 @@
+    for (String column : schemaUpdateActionMap.keySet()) {
+      // this method updates the metadataProperties
+      createColumnWithDefaultValue(column, newMetadataProperties, tmpIndexDir);
+    }
+    // update the actual segment
+    Writer segmentWriter = segmentDirectory.createWriter();
+    List<String> newDimensionColumns = new ArrayList<>();
+    List<String> newMetricColumns = new ArrayList<>();
+    for (String column : schemaUpdateActionMap.keySet()) {
+      Action action = schemaUpdateActionMap.get(column);
+      if (action == Action.ADD_DIMENSION) {
+        newDimensionColumns.add(column);
+      } else if (action == Action.ADD_METRIC) {
+        newMetricColumns.add(column);
+      }
+      if (action == Action.UPDATE_DIMENSION_DEFAULT_VALUE
+          || action == Action.UPDATE_METRIC_DEFAULT_VALUE) {
+        if (segmentWriter.isIndexRemovalSupported()) {
+          segmentWriter.removeIndex(column, ColumnIndexType.DICTIONARY);
+          segmentWriter.removeIndex(column, ColumnIndexType.FORWARD_INDEX);
+        } else {
+          LOGGER.warn("Updating default value is not supported in {} segment version",
+              metadata.getVersion());
+        }
+      }
+      // add dictionary
+      String dictionaryFileName = metadata.getDictionaryFileName(column, metadata.getVersion());
+      File dictionaryFile = new File(tmpIndexDir, dictionaryFileName);
+      PinotDataBuffer dictionaryBuffer = segmentWriter.newIndexFor(column,
+          ColumnIndexType.DICTIONARY, (int) dictionaryFile.length());
+      dictionaryBuffer.readFrom(dictionaryFile);
+      dictionaryBuffer.close();
+      // add forward index
+      String fwdIndexFileName = metadata.getForwardIndexFileName(column, metadata.getVersion());
+      File fwdIndexFile = new File(tmpIndexDir, fwdIndexFileName);
+      PinotDataBuffer fwdIndexBuffer = segmentWriter.newIndexFor(column,
+          ColumnIndexType.FORWARD_INDEX, (int) fwdIndexFile.length());
+      fwdIndexBuffer.readFrom(fwdIndexFile);
+      fwdIndexBuffer.close();
+    }
+    segmentWriter.close();
+    if (newDimensionColumns.size() > 0) {
+      List<String> list =
+          newMetadataProperties.getList(V1Constants.MetadataKeys.Segment.DIMENSIONS);
+      list.addAll(newDimensionColumns);
+      newMetadataProperties.setProperty(V1Constants.MetadataKeys.Segment.DIMENSIONS, list);
+    }
+    if (newMetricColumns.size() > 0) {
+      List<String> list = newMetadataProperties.getList(V1Constants.MetadataKeys.Segment.METRICS);
+      list.addAll(newMetricColumns);
+      newMetadataProperties.setProperty(V1Constants.MetadataKeys.Segment.METRICS, list);
+    }
+    // create a back up.
+    File backupMetadataFile =
+        new File(indexDir, V1Constants.MetadataKeys.METADATA_FILE_NAME + ".bak");
+    File origMetadataFile = new File(indexDir, V1Constants.MetadataKeys.METADATA_FILE_NAME);
+    // If the back up file already exists, don't create it again
```

Comment:
this is back up of original segment metadata file.  Once the back up is created we might have changed the metadata file and then failed. If we create the copy the file again then we will copy the new modified file and will lose the original metadata file.  What we could do is in the beginning if there was metadata.properties.bak file copy that to metadata.properties and remove .bak file. Then we can create bak up file always.